### ISSUE TITLE
[css-content] Implementing content:none on elements is not web-compatible

### ISSUE THREAD

**Issue by MatsPalmgren** [w3c/csswg-drafts#6503]
The [current spec](https://drafts.csswg.org/css-content/#valdef-content-none) says:
> `none`
>    On elements, this inhibits the children of the element from being rendered as children of this element, as if the element was empty.

We implemented that in Gecko ([bug 1699964](https://bugzilla.mozilla.org/show_bug.cgi?id=1699964)) but sadly it broke a bunch of sites so we had to disable it for now ([bug 1719239](https://bugzilla.mozilla.org/show_bug.cgi?id=1719239)).

So the spec needs to say that `none` has no effect on elements (what all UAs currently implement AFAIK).
If we want to retain this feature, then we need a new keyword for it (`children-none` or whatever).

CC @emilio

**Comment by dholbert** [comment]
Just to clarify what "having no effect" is going to mean here:  it looks like in both Firefox and Chromium, `content:none` just computes to `normal`. Is that what we want to spec here?  (i.e. `none` is an alias for `normal` which computes to `normal`)  Or is this going to be something more subtle?

(I tested WebKit as well, and their getComputedStyle implementation seems to just return the empty string for the computed value of `content`, so I'm not as sure what's going on there.)

**Comment by MatsPalmgren** [comment]
Gecko _resolves_ `none` to `normal` for elements but still _computes_ it to `none`, i.e. it inherits as `none`.  For example:
```html
<style>
li { content: none }
li::marker { content: inherit }
</style>

<ul><li>x</ul>

<script>
let li = document.querySelector('li');
li.textContent = getComputedStyle(li).content + ' ' +
                 getComputedStyle(li, '::marker').content;
</script>
```

Firefox: shows "normal none", and **no** bullet since `::marker` inherited `none` (note that we did this even before implementing `content:none` for elements so there was no change in this regard from those patches)
Chrome: shows "normal normal" , and a bullet

I think we need to sort this out while we're here since it might be a web-compat issue too...
See also similar issues: #1757 #821 #4632

**Comment by ellenmartin22** [comment]
> The [current spec](https://drafts.csswg.org/css-content/#valdef-content-none) says:
> 
> > `none`
> 
> >    On elements, this inhibits the children of the element from being rendered as children of this element, as if the element was empty.
> 
> 
> 
> We implemented that in Gecko ([bug 1699964](https://bugzilla.mozilla.org/show_bug.cgi?id=1699964)) but sadly it broke a bunch of sites so we had to disable it for now ([bug 1719239](https://bugzilla.mozilla.org/show_bug.cgi?id=1719239)).
> 
> 
> 
> So the spec needs to say that `none` has no effect on elements (what all UAs currently implement AFAIK).
> 
> If we want to retain this feature, then we need a new keyword for it (`children-none` or whatever).
> 
> 
> 
> CC @emilio

**Comment by dholbert** [comment]
[note: to reduce confusion/noise, I've just deleted a drive-by comment which was just a copypaste of the initial comment here with nothing additional added.  The poster seems to have posted similar comments in other repos as well; not sure if they're a bot or just testing out github, but in any case it wasn't adding information here.]

**Comment by frivoal** [comment]
Agenda+ to decide if we want to make `none` compute to `normal` on regular elements, or keep them as distinct values with identical behavior.

**Comment by Loirooriol** [comment]
That's correct, they are different computed values in Blink, but I don't think that's observable other than in `::marker`, since `content: inherit` doesn't work, and `getComputedStyle` lies.

Regarding WebKit, note that `::marker` doesn't obey `content` at all. But anyways, `normal` and `none` are also different computed values,  since `none` sets the `hasEffectiveContentNone()` flag. Like in Blink, `content: inherit` doesn't work. But WebKit has something interesting: `getComputedStyle` doesn't lie, i.e. it can be `normal` or `none`, whatever you specified. The only exception is that `content: normal` resolves to `none` on `::before` and `::after`.

So this seems to imply that it's web compatible if elements keep `content: none` as-is in `getComputedStyle`. Though it was a [recent change](https://bugs.webkit.org/show_bug.cgi?id=235222), so maybe there hasn't been enough time to encounter compat problems.

**Comment by dholbert** [comment]
Thanks. It sounds like Gecko/Blink/WebKit all represent 'normal' and 'none' as distinct computed values.  So in response to @frivoal 's agenda+ comment...
> Agenda+ to decide if we want to make none compute to normal on regular elements, or keep them as distinct values with identical behavior

Of these two options^, we probably want the latter option (distinct computed values), so as not to unnecessarily force all engines to change (unless there's good reason to do so).

Having said that: there's some additional magic in how the value is used/not-used in each engine, so I'm not sure what makes to spec about that.

**Comment by dholbert** [comment]
So I think the only question here is whether we want to update the resolution above (`spec content:none as having no effect on elements`) to additionally spec that `content:none` resolves to `normal` as part of having-no-effect.

CC @graouts who implemented the recent WebKit commit that keeps it distinct (according to @Loirooriol above), in case they have opinions here, since they are alone on that for the moment (but that is nice & straightforward if it's web-compatible).

**Comment by dholbert** [comment]
Summarizing the CSSWG discussion:

We'd like to circle back to this after https://bugs.webkit.org/show_bug.cgi?id=235222 has gotten some more testing. If I'm understanding @Loirooriol's summary correctly, that change makes WebKit's `getComputedStyle` truthfully report the actual computed of `content`, without any "resolve-none-to-normal" magic. (Though `none` still behaves as `normal` in many situations; i.e. the used value of `none` may still be `normal`.)

If that seems to be web-compatible & WebKit is able to successfully ship that change, then it seems like we should spec that behavior and change Gecko/Blink to match, to remove unnecessary magic & make `getComputedStyle` a bit closer to reporting the actual computed style (rather than the used style).

**Comment by SebastianZ** [comment]
> Though `none` still behaves as `normal` in many situations; i.e. the used value of `none` may still be `normal`.

This is the important point here. For elements, the used value needs to be `normal`. The examples of broken pages linked to from the [Mozilla bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1719239) all had in common that they accidentally hide normal elements with `content: none` while they only wanted to empty the pseudo-elements `::before` and `::after` or not empty the content at all.

Sebastian

Sebastian

**Comment by Loirooriol** [comment]
@SebastianZ The resolution was making `none` compute as-is. On elements, it will still behave like `normal`.
